pkg/controllers/complianceapi/reconciler.go

```diff
@@ -0,0 +1,109 @@
+package complianceapi
+
+import (
+	"context"
+	"errors"
+
+	k8sdepwatches "github.com/stolostron/kubernetes-dependency-watches/client"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	ServiceName  string = "governance-policy-compliance-history-api"
+	RouteName    string = ServiceName
+	DBSecretName string = "governance-policy-database"
+)
+
+var RouteGVR = schema.GroupVersionResource{Group: "route.openshift.io", Version: "v1", Resource: "routes"}
+
+type ComplianceDBSecretReconciler struct {
+	DynamicWatcher k8sdepwatches.DynamicWatcher
+	DynamicClient  *dynamic.DynamicClient
+}
+
+// Reconcile watches the governance-policy-database secret in the controller namespace. On changes it'll handle
+// the Kubernetes objects to expose the compliance history API.
+func (r *ComplianceDBSecretReconciler) Reconcile(
+	ctx context.Context, watcher k8sdepwatches.ObjectIdentifier,
+) (ctrl.Result, error) {
+	// Everything that is watched is in the same namespace
+	ns := watcher.Namespace
+
+	secret, err := r.DynamicWatcher.GetFromCache(
+		schema.GroupVersionKind{Version: "v1", Kind: "Secret"}, ns, DBSecretName,
+	)
+	if err != nil && !errors.Is(err, k8sdepwatches.ErrNoCacheEntry) {
+		return ctrl.Result{}, nil
+	}
+
+	if secret == nil {
+		klog.V(2).Infof(
+			"The Secret %s is not present. Verifying that the Route of %s is deleted.", DBSecretName, RouteName,
+		)
+
+		err = r.DynamicClient.Resource(RouteGVR).Namespace(ns).Delete(ctx, RouteName, metav1.DeleteOptions{})
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				return ctrl.Result{}, nil
+			}
+
+			klog.Errorf("Failed to delete the compliance history API Route of %s: %v", RouteName, err)
+
+			return ctrl.Result{}, err
+		}
+
+		klog.Infof("Deleted the compliance history API Route of %s", RouteName)
+
+		return ctrl.Result{}, nil
+	}
+
+	klog.V(2).Infof(
+		"The Secret %s is present. Verifying that the compliance history API Route of %s is present.",
+		DBSecretName, RouteName,
+	)
+
+	route := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "route.openshift.io/v1",
+			"kind":       "Route",
+			"metadata": map[string]interface{}{
+				"name":      RouteName,
+				"namespace": ns,
+			},
+			"spec": map[string]interface{}{
+				"port": map[string]interface{}{
+					"targetPort": "compliance-history-api",
```

Comment:
I noticed the port was set to this in the test, but I see it's here too.  Is that right?

Reply:
Yeah, it's the port name that I'm using. For example:
https://github.com/open-cluster-management-io/governance-policy-addon-controller/blob/3699932d1491b5ddb9ab40709cd8e96eb1282785/pkg/addon/policyframework/manifests/managedclusterchart/templates/service.yaml#L23